pandas/io/pytables.py

```diff
@@ -331,11 +331,20 @@ def read_hdf(path_or_buf, key=None, **kwargs):
 
     try:
         if key is None:
-            keys = store.keys()
-            if len(keys) != 1:
-                raise ValueError('key must be provided when HDF file contains '
-                                 'multiple datasets.')
-            key = keys[0]
+            groups = store.groups()
+            if len(groups) == 0:
+                raise ValueError('No dataset in HDF file.')
```

Comment:
Not sure that `ValueError` is the right exception here, but at least it's the same type as the exception raised by 0.18.1. (Although the message for 0.18.1 in this case is "key must be provided when HDF file contains multiple datasets.", which is a bit confusing.) And by the way, the exception raised when trying to do `pd.read_hdf('empty.h5', 'some_key')` is (sensibly) "KeyError: 'No object named some_key in the file'". But raising KeyError for the case where key=None and we are trying to automatically figure out the (single, valid) key in the file seems wrong to me. Let me know if you can think of a better exception (or set of exceptions) for this.